Add adaptive brightness functionality

isoamyl50 opened the pull request:
This adds the adaptive brightness functionality, which polls the built-in Retina display's brightness then update the touch bar's brightness value based on it.   

It includes a lookup table as an array, generated by a [python script](https://cdn.discordapp.com/attachments/926106288960729148/1183700880835477534/generate_lookup.py) using the function  

$$f(x) = \sqrt{x} \cdot 11.302676283 + 1$$

with x is the display's brightness value (range from 0 to 509), and f(x) is the touch bar's brightness value (range from 1 to 255).  

I chose this function over the linear function `x / 2` because using this linear function makes the touch bar look too dim at low brightness.  

I put the lookup table in a separate source file, with the purpose of not to make `backlight.rs` look horrible. Maybe executing the function to generate the lookup table when the program starts instead of hard-coding it is better, and I will update it to this way if you guys want it.

WhatAmISupposedToPutHere commented on src/backlight.rs:
```diff
@@ -33,6 +33,16 @@ fn find_backlight() -> Result<PathBuf> {
+            return Ok(entry.path());
+        }
+    }
+    Err(anyhow!("No backlight device found"))
```
Please change this to mention that the missing backlight device is the primary display one

WhatAmISupposedToPutHere commented on src/backlight.rs:
```diff
@@ -80,7 +93,7 @@ impl BacklightManager {
             0
         } else if since_last_active < BRIGHTNESS_DIM_TIMEOUT as u64 {
-            DEFAULT_BRIGHTNESS
+            BRIGHTNESS_LOOKUP_TABLE[read_attr(&self.display_bl_path, "brightness") as usize]
```
I do not think that it is worth it to pre-calculate the lookup table at all, we are running this on a powerful laptop gpu, not on a microcontroller, floating point math is cheap. And the time spent will be dominated by the kernel doing it's own things anyway.

isoamyl50 replied:
So, we should not hard-code the lookup table, but rather generate it at once on program start? Or you mean that we should calculate the brightness value for the touch bar right after we poll the primary display's brightness?

WhatAmISupposedToPutHere reviewed (changes requested):
I like this approach, however there are some issues. You can either fix those, or i will do it myself when i get around to merging it. (I want to also make it instantly respond to brightness changes on primary display)

isoamyl50 commented:
Things that I updated in the last commit:
- Removed the hard-coded lookup table and instead generate it at program start.
- Update the error message when backlight devices not found.
- Update that line of code in find_display_backlight() from `contains` to `eq`uals (because we know exactly the name of the device is `apple-panel-bl`!).

WhatAmISupposedToPutHere commented:
Merged manually.

isoamyl50 commented:
As a side note, remove line 15 from `backlight.rs`:
```rust
const LOOKUP_TABLE_SIZE: usize = MAX_DISPLAY_BRIGHTNESS as usize + 1;
```
Because we do not use the lookup table.  

---

Also, make the brightness level configurable (via a percentage entry in `config.toml`) for the people who do not want their Touch Bar to update its brightness along with the main display. I am a new Rustacean and not comfortable enough for doing this task. Maybe I am going to open an issue addressing this so that some Rust veterans can come and contribute. <3
